File: backend/core/flower.py
"""Flower mode -- walk a small circle around each waypoint in turn.

Pikmin Bloom plants flowers along the walked trail, so circling a spot
covers an area instead of a single line. For every spot the handler:

1. gets there (walks the routed leg, or teleports),
2. waits ``wait_before_s``,
3. walks from the centre out to the circle, around it ``laps`` times
   (a fractional lap stops part-way round), and back to the centre,
4. waits ``wait_after_s`` (plus an optional random pause).

The whole spot sequence repeats ``rounds`` times (``None`` = until the
user stops).
"""

from __future__ import annotations

import logging
import math
from typing import TYPE_CHECKING, Literal

from config import DEFAULT_PAUSE_MAX, DEFAULT_PAUSE_MIN
from core.handler_common import (
    emit_route_path,
    fetch_route_coords,
    finish_mode,
    pause_with_countdown,
    random_pause_seconds,
)
from core.lap_limit import record_lap_and_check_limit
from models.schemas import Coordinate, MovementMode, SimulationState, osrm_profile_for
from utils.geo import haversine_m

if TYPE_CHECKING:
    from core.simulation_engine import SimulationEngine
    from services.cooldown import CooldownTimer

logger = logging.getLogger(__name__)

TransferMode = Literal["walk", "teleport"]

# Metres per degree of latitude for the local equirectangular offset.
# Accurate to well under 1 % at the tens-of-metres radii this mode uses.
_METERS_PER_DEG_LAT = 111_320.0
# Keeps the longitude scale finite right at the poles.
_MIN_COS_LAT = 1e-6
# Closer than this to a spot counts as already there (no transfer leg).
_ARRIVED_THRESHOLD_M = 5.0
# Angular slack when deciding whether a partial lap ends on a vertex.
_ANGLE_EPS = 1e-9
_MIN_ROUTE_COORDS = 2
_PAUSE_SOURCE = "flower"


def offset_point(center: Coordinate, radius_m: float, angle_rad: float) -> Coordinate:
    """Point ``radius_m`` from ``center`` at ``angle_rad`` (0 = north,
    increasing clockwise through east)."""
    cos_lat = max(math.cos(math.radians(center.lat)), _MIN_COS_LAT)
    dlat = radius_m * math.cos(angle_rad) / _METERS_PER_DEG_LAT
    dlng = radius_m * math.sin(angle_rad) / (_METERS_PER_DEG_LAT * cos_lat)
    return Coordinate(lat=center.lat + dlat, lng=center.lng + dlng)


def circle_points(center: Coordinate, radius_m: float, segments: int) -> list[Coordinate]:
    """``segments`` evenly spaced vertices on the circle, starting north."""
    step = 2.0 * math.pi / segments
    return [offset_point(center, radius_m, k * step) for k in range(segments)]


def circle_path(
    center: Coordinate, radius_m: float, segments: int, laps: float,
) -> list[Coordinate]:
    """The walked path for one spot: centre → circle → ``laps`` times
    around → centre.

    ``laps`` may be fractional. The path follows the polygon vertices
    and, when the lap fraction doesn't land on a vertex, ends on the
    exact angle ``laps · 2π`` so half a lap really is half the circle.
    """
    step = 2.0 * math.pi / segments
    total_angle = laps * 2.0 * math.pi
    full_steps = int(math.floor(total_angle / step + _ANGLE_EPS))
    path = [center]
    path.extend(offset_point(center, radius_m, k * step) for k in range(full_steps + 1))
    if total_angle - full_steps * step > _ANGLE_EPS:
        path.append(offset_point(center, radius_m, total_angle))
    path.append(center)
    return path


def path_length_m(path: list[Coordinate]) -> float:
    """Sum of great-circle distances along ``path``."""
    return sum(
        haversine_m(a.lat, a.lng, b.lat, b.lng)
        for a, b in zip(path, path[1:])
    )


class FlowerHandler:
    """Visits each waypoint and walks a circle around it."""

    def __init__(self, engine: "SimulationEngine") -> None:
        self.engine = engine

    async def start(
        self,
        waypoints: list[Coordinate],
        mode: MovementMode,
        *,
        radius_m: float,
        segments: int,
        laps: float,
        rounds: int | None,
        wait_before_s: float = 0.0,
        wait_after_s: float = 0.0,
        transfer: TransferMode = "walk",
        speed_kmh: float | None = None,
        speed_min_kmh: float | None = None,
        speed_max_kmh: float | None = None,
        pause_enabled: bool = False,
        pause_min: float = DEFAULT_PAUSE_MIN,
        pause_max: float = DEFAULT_PAUSE_MAX,
        straight_line: bool = False,
        cooldown: "CooldownTimer | None" = None,
    ) -> None:
        """Run flower mode until every round finishes or the user stops.

        ``cooldown`` is honoured only for teleport transfers: each hop
        first waits out an active cooldown, then starts a new one for the
        jump distance. Callers pass ``None`` where cooldown is bypassed
        (group mode).
        """
        engine = self.engine

        if not waypoints:
            raise ValueError("At least 1 waypoint is required for flower mode")
        if transfer == "walk" and engine.current_position is None:
            raise RuntimeError(
                "Cannot start flower mode: no current position. Teleport first."
            )

        profile_name = mode.value
        osrm_profile = osrm_profile_for(mode)
        total_spots = len(waypoints)

        def pick_profile() -> dict:
            return engine.pick_speed_profile(
                profile_name, speed_kmh, speed_min_kmh, speed_max_kmh,
            )

        engine.state = SimulationState.FLOWER
        engine.lap_count = 0
        engine.segment_index = 0
        engine.distance_traveled = 0.0
        engine._route_offset_remaining = 0.0
        # Named-waypoint highlighting is emitted explicitly on arrival;
        # the movement loop's pass detection would misfire on the circles.
        engine._user_waypoints = []
        engine._user_waypoint_next = 0

        await engine._emit("state_change", {
            "state": engine.state.value,
            "waypoints": [{"lat": wp.lat, "lng": wp.lng} for wp in waypoints],
        })

        logger.info(
            "Flower started: %d spots, r=%.0fm, seg=%d, laps=%.1f, rounds=%s, "
            "wait=%.0f/%.0fs, transfer=%s [%s]",
            total_spots, radius_m, segments, laps, rounds if rounds else "∞",
            wait_before_s, wait_after_s, transfer, profile_name,
        )

        while not engine._stop_event.is_set():
            stopped = False
            for idx, spot in enumerate(waypoints):
                engine.segment_index = idx
                if await self._visit_spot(
                    idx, spot, total_spots,
                    radius_m=radius_m, segments=segments, laps=laps,
                    wait_before_s=wait_before_s, wait_after_s=wait_after_s,
                    transfer=transfer, osrm_profile=osrm_profile,
                    straight_line=straight_line, pick_profile=pick_profile,
                    pause_enabled=pause_enabled, pause_min=pause_min,
                    pause_max=pause_max, cooldown=cooldown,
                ):
                    stopped = True
                    break
            if stopped or engine._stop_event.is_set():
                break
            if await record_lap_and_check_limit(
                engine, rounds, kind="Flower", logger=logger,
            ):
                break

        engine._route_offset_remaining = 0.0
        await finish_mode(engine, (SimulationState.FLOWER,))
        logger.info("Flower finished after %d rounds", engine.lap_count)

    async def _visit_spot(
        self,
        idx: int,
        spot: Coordinate,
        total_spots: int,
        *,
        radius_m: float,
        segments: int,
        laps: float,
        wait_before_s: float,
        wait_after_s: float,
        transfer: TransferMode,
        osrm_profile: str,
        straight_line: bool,
        pick_profile,
        pause_enabled: bool,
        pause_min: float,
        pause_max: float,
        cooldown: "CooldownTimer | None",
    ) -> bool:
        """Run the full per-spot sequence. Returns True when stopped."""
        engine = self.engine

        if transfer == "teleport":
            if await self._teleport_to(spot, cooldown):
                return True
        else:
            await self._walk_to(spot, osrm_profile, straight_line, pick_profile)
        if engine._stop_event.is_set():
            return True

        await engine._emit("waypoint_progress", {
            "current_index": idx,
            "next_index": min(idx + 1, total_spots - 1),
            "total": total_spots,
        })

        if wait_before_s > 0 and await pause_with_countdown(
            engine, wait_before_s, _PAUSE_SOURCE,
        ):
            return True

        path = circle_path(spot, radius_m, segments, laps)
        await emit_route_path(engine, path)
        await engine._move_along_route(path, pick_profile())
        if engine._stop_event.is_set():
            return True

        after = wait_after_s
        if pause_enabled:
            after += random_pause_seconds(pause_min, pause_max)
        if after > 0 and await pause_with_countdown(engine, after, _PAUSE_SOURCE):
            return True
        return False

    async def _walk_to(
        self,
        spot: Coordinate,
        osrm_profile: str,
        straight_line: bool,
        pick_profile,
    ) -> None:
        """Walk the routed leg from the current position to ``spot``.

        Route planning failures (``RouteUnavailableError``) propagate so
        the run aborts with a visible error instead of silently jumping.
        """
        engine = self.engine
        here = engine.current_position
        if here is None:
            raise RuntimeError("Cannot walk to the next flower spot: no current position")
        if haversine_m(here.lat, here.lng, spot.lat, spot.lng) <= _ARRIVED_THRESHOLD_M:
            return
        coords, _route_data = await fetch_route_coords(
            engine.route_service,
            here.lat, here.lng,
            spot.lat, spot.lng,
            profile=osrm_profile,
            straight=straight_line,
        )
        if len(coords) < _MIN_ROUTE_COORDS:
            return
        await emit_route_path(engine, coords)
        await engine._move_along_route(coords, pick_profile())

    async def _teleport_to(
        self, spot: Coordinate, cooldown: "CooldownTimer | None",
    ) -> bool:
        """Jump to ``spot``, respecting the cooldown. Returns True when stopped.

        Deliberately does not call ``engine.teleport()``: that stops the
        active task, which is this handler.
        """
        engine = self.engine
        # A jump is a single push, so it would bypass the movement loop's
        # pause check — honour a pause here instead.
        await engine._pause_event.wait()
        if engine._stop_event.is_set():
            return True
        here = engine.current_position
        if here is not None and haversine_m(
            here.lat, here.lng, spot.lat, spot.lng,
        ) <= _ARRIVED_THRESHOLD_M:
            return False

        if cooldown is not None and cooldown.enabled and cooldown.is_active:
            remaining = float(cooldown.remaining)
            if remaining > 0 and await pause_with_countdown(
                engine, remaining, _PAUSE_SOURCE,
            ):
                return True

        await engine._set_position(spot.lat, spot.lng)
        await engine._emit("teleport", {"lat": spot.lat, "lng": spot.lng})
        await engine._emit("position_update", {"lat": spot.lat, "lng": spot.lng})

        if cooldown is not None and cooldown.enabled and here is not None:
            await cooldown.start(here.lat, here.lng, spot.lat, spot.lng)
        return False

File: frontend/src/lib/flower.ts
// Flower mode settings + plan estimate. Pure functions — behaviour is
// pinned by ./flower.test.ts. Bounds mirror backend `FlowerRequest`.
import { haversineM } from './geo'

type LatLng = { lat: number; lng: number }

export type FlowerTransfer = 'walk' | 'teleport'

export interface FlowerSettings {
  radiusM: number
  /** Vertices per circle. */
  segments: number
  /** Laps around each spot, in half-lap steps. */
  laps: number
  /** Passes over the spot list. null = until stopped. */
  rounds: number | null
  waitBeforeS: number
  waitAfterS: number
  transfer: FlowerTransfer
}

export const FLOWER_LIMITS = {
  radiusM: { min: 5, max: 100, step: 5 },
  segments: { min: 6, max: 24, step: 2 },
  laps: { min: 0.5, max: 10, step: 0.5 },
  rounds: { min: 1, max: 99 },
  waitS: { min: 0, max: 600 },
} as const

/** Stepper ladder for the wait controls (seconds). */
export const FLOWER_WAIT_STEPS_S = [0, 5, 10, 15, 30, 60, 120, 180, 300, 600] as const

export const DEFAULT_FLOWER_SETTINGS: FlowerSettings = {
  radiusM: 20,
  segments: 12,
  laps: 1,
  rounds: 1,
  waitBeforeS: 0,
  waitAfterS: 0,
  transfer: 'walk',
}

function clamp(v: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, v))
}

function finite(v: unknown): number | null {
  return typeof v === 'number' && Number.isFinite(v) ? v : null
}

/** Coerce anything (e.g. a stored JSON blob) into valid settings. */
export function sanitizeFlowerSettings(raw: unknown): FlowerSettings {
  const o = (raw != null && typeof raw === 'object' ? raw : {}) as Record<string, unknown>
  const d = DEFAULT_FLOWER_SETTINGS
  const L = FLOWER_LIMITS
  const radius = finite(o.radiusM)
  const segments = finite(o.segments)
  const laps = finite(o.laps)
  const rounds = finite(o.rounds)
  const before = finite(o.waitBeforeS)
  const after = finite(o.waitAfterS)
  return {
    radiusM: radius == null ? d.radiusM : clamp(radius, L.radiusM.min, L.radiusM.max),
    segments: segments == null ? d.segments : clamp(Math.round(segments), L.segments.min, L.segments.max),
    laps: laps == null
      ? d.laps
      : clamp(Math.round(laps / L.laps.step) * L.laps.step, L.laps.min, L.laps.max),
    rounds: o.rounds === null
      ? null
      : rounds == null ? d.rounds : clamp(Math.round(rounds), L.rounds.min, L.rounds.max),
    waitBeforeS: before == null ? d.waitBeforeS : clamp(before, L.waitS.min, L.waitS.max),
    waitAfterS: after == null ? d.waitAfterS : clamp(after, L.waitS.min, L.waitS.max),
    transfer: o.transfer === 'teleport' ? 'teleport' : 'walk',
  }
}

/** Next value on the wait ladder in direction `dir` (+1 / -1). */
export function stepFlowerWait(current: number, dir: 1 | -1): number {
  const steps = FLOWER_WAIT_STEPS_S
  if (dir > 0) return steps.find((s) => s > current) ?? steps[steps.length - 1]
  for (let i = steps.length - 1; i >= 0; i--) {
    if (steps[i] < current) return steps[i]
  }
  return steps[0]
}

/** Next rounds value: 1…99, then unlimited (null); down from unlimited
 *  lands on 99. */
export function stepFlowerRounds(current: number | null, dir: 1 | -1): number | null {
  const { min, max } = FLOWER_LIMITS.rounds
  if (current === null) return dir > 0 ? null : max
  if (dir > 0) return current >= max ? null : current + 1
  return Math.max(min, current - 1)
}

/** Length of one spot's walked path: centre → circle → `laps` around the
 *  polygon (partial final chord for a fractional lap) → centre. Mirrors
 *  backend `core.flower.circle_path`. */
export function flowerCirclePathM(radiusM: number, segments: number, laps: number): number {
  const step = (2 * Math.PI) / segments
  const total = laps * 2 * Math.PI
  const fullSteps = Math.floor(total / step + 1e-9)
  const chord = 2 * radiusM * Math.sin(step / 2)
  const rem = total - fullSteps * step
  const partial = rem > 1e-9 ? 2 * radiusM * Math.sin(rem / 2) : 0
  return 2 * radiusM + fullSteps * chord + partial
}

export interface FlowerPlan {
  /** Walking distance of the first round (includes reaching spot 1). */
  firstRoundM: number
  /** Walking distance of every later round (includes last → first). */
  laterRoundM: number
  /** Total waiting per round. */
  roundWaitS: number
  rounds: number | null
}

export function estimateFlowerPlan(
  spots: LatLng[],
  settings: FlowerSettings,
  start: LatLng | null,
): FlowerPlan {
  const n = spots.length
  const circles = n * flowerCirclePathM(settings.radiusM, settings.segments, settings.laps)
  let chain = 0
  let toFirst = 0
  let wrap = 0
  if (settings.transfer === 'walk' && n > 0) {
    for (let i = 0; i + 1 < n; i++) chain += haversineM(spots[i], spots[i + 1])
    toFirst = start ? haversineM(start, spots[0]) : 0
    wrap = haversineM(spots[n - 1], spots[0])
  }
  return {
    firstRoundM: n === 0 ? 0 : circles + chain + toFirst,
    laterRoundM: n === 0 ? 0 : circles + chain + wrap,
    roundWaitS: n * (settings.waitBeforeS + settings.waitAfterS),
    rounds: settings.rounds,
  }
}

/** Total walking distance; null when rounds are unlimited. */
export function flowerPlanDistanceM(plan: FlowerPlan): number | null {
  if (plan.rounds === null) return null
  return plan.firstRoundM + (plan.rounds - 1) * plan.laterRoundM
}

/** Estimated run time in seconds; null when rounds are unlimited or the
 *  speed is not positive. */
export function flowerPlanSeconds(plan: FlowerPlan, speedKmh: number): number | null {
  const dist = flowerPlanDistanceM(plan)
  if (dist === null || speedKmh <= 0 || plan.rounds === null) return null
  return dist / (speedKmh / 3.6) + plan.rounds * plan.roundWaitS
}
